lib/go/test/flow_epoch_test.go

```diff
@@ -0,0 +1,795 @@
+		assert.Equal(t, cadence.NewUInt64(67), result)
+
+		// DKG Contract Checks
+		result = executeScriptAndCheck(t, b, templates.GenerateGetDKGEnabledScript(env), nil)
+		assert.Equal(t, cadence.NewBool(true), result)
+
+		result = executeScriptAndCheck(t, b, templates.GenerateGetDKGNodeIsRegisteredScript(env), [][]byte{jsoncdc.MustEncode(cadence.String(ids[1]))})
+		assert.Equal(t, cadence.NewBool(true), result)
+
+		result = executeScriptAndCheck(t, b, templates.GenerateGetConsensusNodesScript(env), nil)
+		assert.Equal(t, cadence.NewArray(dkgIDs), result)
+
+		result = executeScriptAndCheck(t, b, templates.GenerateGetDKGFinalSubmissionsScript(env), nil)
+		assert.Equal(t, cadence.NewArray(make([]cadence.Value, 0)), result)
+
+		result = executeScriptAndCheck(t, b, templates.GenerateGetDKGCompletedScript(env), nil)
+		assert.Equal(t, cadence.NewBool(false), result)
+
+	})
+
+	t.Run("Should not be able to advance to epoch committed or end epoch during epoch committed if nothing has happened", func(t *testing.T) {
+		// try to advance to the epoch committed phase
+		// will not panic, but no state has changed
+		advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "EPOCHCOMMITTED", false)
+
+		verifyConfigMetadata(t, b, env,
+			ConfigMetadata{
+				currentEpochCounter:      startEpochCounter,
+				proposedEpochCounter:     startEpochCounter + 1,
+				currentEpochPhase:        1,
+				numViewsInEpoch:          numEpochViews,
+				numViewsInStakingAuction: numStakingViews,
+				numViewsInDKGPhase:       numDKGViews,
+				numCollectorClusters:     numClusters})
+
+		// try to advance to the end epoch phase
+		// will fail
+		advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "ENDEPOCH", true)
+	})
+
+}
+
+func TestEpochQCDKGNodeRegistration(t *testing.T) {
+	b, accountKeys, env := newTestSetup(t)
+
+	// Create new keys for the epoch account
+	idTableAccountKey, IDTableSigner := accountKeys.NewWithSigner()
+
+	// Deploys the staking contract, qc, dkg, and epoch lifecycle contract
+	// staking contract is deployed with default values (1.25M rewards, 8% cut)
+	idTableAddress := initializeAllEpochContracts(t, b, idTableAccountKey, IDTableSigner, &env,
+		0,             // start epoch counter
+		70,            // num views per epoch
+		50,            // num views for staking auction
+		2,             // num views for DKG phase
+		4,             // num collector clusters
+		"lolsoRandom") // random source
+
+	// create new user accounts, mint tokens for them, and register them for staking
+	addresses, signers := registerAndMintManyAccounts(t, b, accountKeys, numEpochAccounts)
+	ids, _, _ := generateNodeIDs(numEpochAccounts)
+	registerNodesForStaking(t, b, env,
+		addresses,
+		signers,
+		ids)
+
+	t.Run("Should not be able to register a QC voter or DKG participant if the staking auction isn't over", func(t *testing.T) {
```

Comment:
I'm not sure this will work with the operator onboarding flow we discussed 🤔. We wanted to:
* submit the staking request
* create the node operator's machine account
* retrieve the DKGParticipant or QCVoter resource and insert it to the machine account

all at the same time (ideally in one transaction) through Flow Port. So we would need to allow retrieving the QC/DKG resources during staking auction, right?

Reply:
We can't allow retrieving the resources during the staking auction because those can only be retrieved by approved nodes which doesn't happen until the end of the staking auction. As long as the node operator controls both accounts, then we should be able to do it.